crates/sui-json-rpc/src/coin_api.rs

```diff
@@ -288,41 +209,706 @@ impl CoinReadApiServer for CoinReadApi {
+        let treasury_cap_object = self
+            .state
+            .find_package_object(&coin_struct.address.into(), TreasuryCap::type_(coin_struct))
+            .await?;
+
+        Ok(TreasuryCap::from_bcs_bytes(
+            treasury_cap_object.data.try_as_move().unwrap().contents(),
+        )?)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use jsonrpsee::types::ErrorObjectOwned;
+    use move_core_types::language_storage::StructTag;
+    use sui_types::balance::Supply;
+    use sui_types::base_types::{ObjectID, SuiAddress};
+    use sui_types::coin::TreasuryCap;
+    use sui_types::id::UID;
+    use sui_types::object::Object;
+    use sui_types::parse_sui_struct_tag;
+
+    fn get_test_owner() -> SuiAddress {
+        SuiAddress::random_for_testing_only()
+    }
+
+    fn get_test_package_id() -> ObjectID {
+        ObjectID::random()
+    }
+
+    fn get_test_coin_type(package_id: ObjectID) -> String {
+        format!("{}::test_coin::TEST_COIN", package_id)
+    }
+
+    fn get_test_treasury_cap_peripherals(
+        package_id: ObjectID,
+    ) -> (String, StructTag, StructTag, TreasuryCap, Object) {
+        let coin_name = get_test_coin_type(package_id);
+        let input_coin_struct = parse_sui_struct_tag(&coin_name).expect("should not fail");
+        let treasury_cap_struct = TreasuryCap::type_(input_coin_struct.clone());
+        let treasury_cap = TreasuryCap {
+            id: UID::new(ObjectID::random()),
+            total_supply: Supply { value: 420 },
+        };
+        let treasury_cap_object =
+            Object::treasury_cap_for_testing(input_coin_struct.clone(), treasury_cap.clone());
+        (
+            coin_name,
+            input_coin_struct,
+            treasury_cap_struct,
+            treasury_cap,
+            treasury_cap_object,
+        )
+    }
+
+    mod get_coins_tests {
+        use super::super::*;
+        use super::*;
+        use jsonrpsee::types::ErrorObjectOwned;
+        use typed_store::TypedStoreError;
+
+        #[tokio::test]
+        async fn test_invalid_coin_type() {
+            let owner = get_test_owner();
+            let coin_type = "0x2::invalid::struct::tag";
+            let mock_internal = MockCoinReadInternalTrait::new();
+            let coin_read_api = CoinReadApi {
+                internal: Arc::new(mock_internal),
+            };
+
+            let response = coin_read_api
+                .get_coins(owner, Some(coin_type.to_string()), None, None)
+                .await;
+
+            assert!(response.is_err());
+            let error_result = response.unwrap_err();
+            let error_object: ErrorObjectOwned = error_result.into();
+            assert_eq!(error_object.code(), -32602);
+            assert_eq!(
+                error_object.message(),
+                "Invalid struct type: 0x2::invalid::struct::tag. Got error: Expected end of token stream. Got: ::"
+            );
```

Comment:
I wonder if `expect_test` might help here, cc @rvantonder

Reply:
Oh neat, so this will automatically update tests if I set `UPDATE_EXPECT=1`?

Reply:
That's right.  I just used it in some of my work recently -- it sometimes requiers a bit of finessing to scrub device-specific paths etc, but after that it really speeds things up!